hs_restclient/__init__.py

```diff
@@ -1 +1,666 @@
+
+        :param hostname: Hostname of the HydroShare server to query
+        :param auth: Concrete instance of AbstractHydroShareAuth (i.e. HydroShareAuthBasic)
+        :param use_https: Boolean, if True, HTTPS will be used
+        :param port: Integer representing the TCP port on which to connect
+            to the HydroShare server
+
+        :raises: HydroShareException if auth is not a known authentication type.
+    """
+
+    _URL_PROTO_WITHOUT_PORT = "{scheme}://{hostname}/hsapi"
+    _URL_PROTO_WITH_PORT = "{scheme}://{hostname}:{port}/hsapi"
+
+    def __init__(self, hostname='www.hydroshare.org', auth=None,
+                 use_https=False, port=None):
+        self.hostname = hostname
+
+        self.session = None
+        self.auth = None
+        if auth:
+            if isinstance(auth, HydroShareAuthBasic):
+                # HTTP basic authentication
+                self.auth = (auth.username, auth.password)
+            else:
+                raise HydroShareException("Unsupported authentication type '{0}'.".format(str(type(auth))))
+
+        if use_https:
+            self.scheme = 'https'
+        else:
+            self.scheme = 'http'
+        if port:
+            self.port = int(port)
+            if self.port < 0 or self.port > 65535:
+                raise HydroShareException("Port number {0} is illegal.".format(self.port))
+            self.url_base = self._URL_PROTO_WITH_PORT.format(scheme=self.scheme,
+                                                             hostname=self.hostname,
+                                                             port=self.port)
+        else:
+            self.url_base = self._URL_PROTO_WITHOUT_PORT.format(scheme=self.scheme,
+                                                                hostname=self.hostname)
+
+        self._initializeSession()
+        self._resource_types = None
+
+    @property
+    def resource_types(self):
+        if self._resource_types is None:
+            self._resource_types = self.getResourceTypes()
+        return self._resource_types
+
+
+    def _initializeSession(self):
+        if self.session:
+            self.session.close()
+        self.session = requests.Session()
+        self.session.auth = self.auth
+
+    def _request(self, method, url, params=None, data=None, files=None, stream=False):
+        r = None
+        try:
+            r = self.session.request(method, url, params=params, data=data, files=files, stream=stream)
+        except requests.ConnectionError:
+            # We might have gotten a connection error because the server we were talking to went down.
+            #  Re-initialize the session and try again
+            self._initializeSession()
+            r = self.session.request(method, url, params=params, data=data, files=files, stream=stream)
+
+        return r
+
+    def _prepareFileForUpload(self, resource_file, resource_filename=None):
+        if type(resource_file) is str:
```

Comment:
what if it is a unicode object? Use `isinstance(resource_file, basestring)`

If you need to support both Python 2 and Python 3 you'll need to add this to the top of the module to define `basestring` on Python 3 where there is no more `str/unicode`

```
try:
    unicode
except NameError: # Python 3
    basestring = str
```